packages/router/router.ts

```diff
@@ -790,12 +887,96 @@ export function createRouter(init: RouterInit): Router {
     cancelledFetcherLoads = [];
   }
 
+  function createBlocker(key: string, fn: ShouldBlockFunction) {
+    let blocker: Blocker = {
+      state: "unblocked",
+      proceed: undefined,
+      reset: undefined,
+      fn,
+    };
+    state.blockers.set(key, blocker);
+    return blocker;
+  }
+
+  function getBlocker(key: string) {
+    return state.blockers.get(key);
+  }
+
+  function deleteBlocker(key: string) {
+    state.blockers.delete(key);
+  }
+
+  function setBlockerState(
+    key: string,
+    state: "blocked",
+    navigate: () => Promise<void>
+  ): Blocker | undefined;
+  function setBlockerState(
+    key: string,
+    state: "unblocked" | "proceeding"
+  ): Blocker | undefined;
+
+  function setBlockerState(
+    key: string,
+    nextState: Blocker["state"],
+    navigate?: () => Promise<void>
+  ) {
+    let blocker = state.blockers.get(key);
+    if (!blocker) return;
+    invariant(
+      nextState === "proceeding" ||
+        nextState === "blocked" ||
+        nextState === "unblocked",
+      `Invalid blocker state: ${nextState}`
+    );
+
+    blocker.state = nextState;
+    if (nextState === "blocked") {
+      blocker.proceed = async () => {
+        invariant(
+          typeof navigate === "function",
+          "Cannot proceed without a navigate function"
+        );
+        setBlockerState(key, "proceeding");
+        await navigate();
+      };
+      blocker.reset = () => {
+        setBlockerState(key, "unblocked");
+      };
+    } else {
+      blocker.proceed = undefined;
+      blocker.reset = undefined;
+    }
+    return blocker;
+  }
+
+  function setBlockerFunction(key: string, fn: Blocker["fn"]) {
+    let blocker = state.blockers.get(key);
+    if (!blocker) return;
+    if (typeof fn !== "function") {
+      warning(
+        false,
+        `A blocker function update was requested with a value that is not a function. This is not allowed, and the blocker's function will not be updated.`
+      );
+      return;
+    }
+    blocker.fn = fn;
+    return blocker;
+  }
+
   // Trigger a navigation event, which can either be a numerical POP or a PUSH
   // replace with an optional submission
   async function navigate(
     to: number | To,
     opts?: RouterNavigateOptions
   ): Promise<void> {
+    for (let [key, blocker] of state.blockers) {
+      if (blocker.state !== "proceeding" && blocker.fn()) {
```

Comment:
Can there ever be multiple `proceeding` blockers?  If so we may need to be key aware here since we only want to skip the blocker function for the specific blocker that called `proceed()`